test/common/buffer/buffer_speed_test.cc

```diff
@@ -24,6 +33,76 @@ static void bufferCreateEmpty(benchmark::State& state) {
 }
 BENCHMARK(bufferCreateEmpty);
 
+// Test add performance of OwnedImpl vs WatermarkBuffer
+static void bufferVsWatermarkBuffer(benchmark::State& state) {
+  const uint64_t length = state.range(0);
+  const uint64_t high_watermark = state.range(1);
+  const bool use_watermark_buffer = (state.range(2) != 0);
+
+  for (auto _ : state) {
+    UNREFERENCED_PARAMETER(_);
+    std::unique_ptr<Buffer::Instance> buffer;
+    if (use_watermark_buffer) {
+      buffer = std::make_unique<Buffer::WatermarkBuffer>([]() {}, []() {}, []() {});
+      buffer->setWatermarks(high_watermark);
+    } else {
+      buffer = std::make_unique<Buffer::OwnedImpl>();
+    }
+
+    for (uint64_t idx = 0; idx < length; ++idx) {
+      buffer->add("a", 1);
+    }
+  }
+  benchmark::DoNotOptimize(length);
+}
+BENCHMARK(bufferVsWatermarkBuffer)
+    ->Args({1024, 0, 0})
+    ->Args({1024, 0, 1})
+    ->Args({1024, 1, 1})
+    ->Args({1024, 1024, 1})
+    ->Args({64 * 1024, 0, 0})
+    ->Args({64 * 1024, 0, 1})
+    ->Args({64 * 1024, 1, 1})
+    ->Args({64 * 1024, 64 * 1024, 1})
+    ->Args({1024 * 1024, 0, 0})
+    ->Args({1024 * 1024, 0, 1})
+    ->Args({1024 * 1024, 1, 1})
+    ->Args({1024 * 1024, 1024 * 1024, 1});
+
+// Measure performance impact of enabling accounts.
+static void bufferAccountUse(benchmark::State& state) {
+  const std::string data(state.range(0), 'a');
+  uint64_t iters = state.range(1);
+  const bool enable_accounts = (state.range(2) != 0);
+
+  auto config = envoy::config::overload::v3::BufferFactoryConfig();
+  config.set_minimum_account_to_track_power_of_two(2);
+  Buffer::WatermarkBufferFactory buffer_factory(config);
+  FakeStreamResetHandler reset_handler;
+
+  for (auto _ : state) {
+    UNREFERENCED_PARAMETER(_);
+    Buffer::OwnedImpl buffer;
+    if (enable_accounts) {
+      auto account = buffer_factory.createAccount(reset_handler);
```

Comment:
We'll want to call `clearDownstream()` on this account when we're finished with it

Reply:
The reset_handler outlives the account so it shouldn't be needed.  I changed the benchmark so clearDownstream() is called.